### PR TITLE
refactor(core/pipeline): Convert PipelineGraph to React

### PR DESCRIPTION
**Pull request by jrsquared**
Convert PipelineGraph to React.

A few things
1. Made worse: Somewhat slower to respond to mouse hovering than before. I optimized some, but there's still definitely room for improvement.
2. Fixed: The graph actually updates when stage status updates.
3. Fixed: Label styles now update appropriately when hovering over the node.

After converting it, I noticed there should probably be two separate higher-order components for pipeline configuration graph and execution graph (lots of `if (props.execution)` littered in the component, but that can be attacked later... which means it never will be.

I'm sure I broke a lot of things that I missed, so hopefully @anotherchrisberry and @danielpeach will catch the rest.

### CODE REVIEW

File: app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx [hunk 1]
```diff
@@ -1,11 +1,464 @@
...
+import './pipelineGraph.less'
 
 export interface IPipelineGraphProps {
+  execution?: IExecution;
+  onNodeClick: (node: IPipelineGraphNode) => void;
   pipeline?: IPipeline;
-  execution: IExecution;
   viewState: IExecutionViewState;
-  onNodeClick: (node: IPipelineNode) => void;
-  shouldValidate?: boolean;
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx]: you sure you want to get rid of this?

File: app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx [hunk 2]
```diff
@@ -1,11 +1,464 @@
...
+  /**
+   * Used to draw inverse bezier curve between stages
+   */
+  private curvedLink(d: any) {
+    const sourceX = d.source.x + this.state.nodeRadius;
+    const targetX = d.target.x - this.state.nodeRadius;
+    return 'M' + sourceX + ',' + d.source.y
+        + 'C' + (sourceX + targetX) / 2 + ',' + d.source.y
+        + ' ' + (sourceX + targetX) / 2 + ',' + d.target.y
+        + ' ' + targetX + ',' + d.target.y;
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx]: Is this easier to read or harder to read as a string template?
```javascript
const curve = (sourceX + targetX) / 2;
return `M${sourceX},${d.source.y}C${curve},${d.source.y} ${curve},${d.target.y} ${targetX},${d.target.y}`;
```

File: app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx [hunk 3]
```diff
@@ -1,11 +1,464 @@
...
+      return node.phase;
+    }
+    return max(node.children.map((n) => this.getLastPhase(n)));
+  }
+
+  private createNodes(props: IPipelineGraphProps): IPipelineGraphNode[] {
+    const { execution, pipeline, viewState } = props;
+    return pipeline ?
+    PipelineGraphService.generateConfigGraph(pipeline, viewState, this.pipelineValidations) :
+    PipelineGraphService.generateExecutionGraph(execution, viewState);
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx]: these two lines should be indented

File: app/scripts/modules/core/src/pipeline/config/pipelineConfigurer.js [hunk 4]
```diff
@@ -35,14 +35,16 @@ module.exports = angular.module('spinnaker.core.pipeline.config.pipelineConfigur
     // For standard pipelines, a 'renderablePipeline' is just the pipeline config.
     // For templated pipelines, a 'renderablePipeline' is the pipeline template plan, and '$scope.pipeline' is the template config.
     $scope.renderablePipeline = $scope.plan || $scope.pipeline;
-
+    // Watch for non-refernce changes to renderablePipline and make them reference changes to make React happy
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/pipeline/config/pipelineConfigurer.js]: (non-reference)

File: app/scripts/modules/core/src/pipeline/config/pipelineConfigurer.js [hunk 5]
```diff
@@ -246,58 +250,70 @@ module.exports = angular.module('spinnaker.core.pipeline.config.pipelineConfigur
       });
     };
 
-    this.navigateToStage = function(index, event) {
+    // Poor react setState
+    this.setViewState = (newViewState) => {
+      const viewState = _.clone($scope.viewState);
+      Object.assign(viewState, newViewState);
+      $scope.$applyAsync(() => $scope.viewState = viewState);
+    };
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/pipeline/config/pipelineConfigurer.js]: nice

File: package.json [hunk 6]
```diff
@@ -69,7 +69,7 @@
     "react-ga": "^2.1.2",
     "react-select": "^1.0.0-rc.5",
     "react-virtualized-select": "^3.1.0",
-    "react2angular": "^1.1.3",
+    "react2angular": "^2.0.0",
```
Comment (anotherchrisberry) [inline package.json]: anything different in 2.0 or just keeping up with dependencies?

File: app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx (hunk 1 above)
Comment (jrsquared) [reply app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx]: Nope, stupid string search for `shouldValidate` not thinking about angular `should-validate`.

File: app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx (hunk 2 above)
Comment (jrsquared) [reply app/scripts/modules/core/src/pipeline/config/graph/PipelineGraph.tsx]: Easier. I'll change it.

File: package.json (hunk 6 above)
Comment (jrsquared) [reply package.json]: They did have a change regarding scope changes that I was hoping would stop me from having to $applyAsync, but I still had to do it. Left the update in anyways.

**Comment by anotherchrisberry** [comment]
LGTM, merge away if you're feeling lucky

**Pull request merged**